### PR TITLE
Fix GAN training by replacing zero_grad logic

### PR DESCRIPTION
**Pull request by aybberk**
Hope this solves the issue https://github.com/williamFalcon/pytorch-lightning/issues/591 without messing with other things. I removed zero_grad logic from optimizer_step to outside since optimizer_step function does not have access to the other optimizers to zero_grad and thought self.zero_grad would not permit calling on_before_zero_grad in the future.

### CODE REVIEW

**Comment by williamFalcon** [comment]
@aybberk actually, i think this makes sense. users shouldn't be worried about gradient behavior when using multiple optimizers

**Comment by aybberk** [comment]
Hi, I was thinking about a more generic way to handle this issue since I suppose this solution may mess with accumulating gradients. After stepping with one optimizer, my intuition says one may want to keep other parameters' gradients that have possibly been accumulated over batches although I'm not 100% sure about that.

So, what I have been thinking to solve this issue is following. We should make sure that only the gradients for the current optimizer's parameters are calculated and updated but it must happen outside of the training loop. The most straightforward way to ensure this is turn off the require_grad flag of all parameters except the ones the current optimizer is responsible for, just before the optimizer_closure function is called in the run_training_batch method. 

One caveat would this solution may be turning off the require_grad flag may reset the accumulated gradients of the parameters, but I have made a mini test for this and can say it does not happen. 

```
# Alternatingly turn off and on the require_grad flag of x
# to see if it prevents the gradient accumulation or not

x = torch.zeros(3, requires_grad = True)
y = x.mean()

for n in range(10):
    if n % 2 == 0:
        x.requires_grad = False
    else:
        x.requires_grad = True
    
    y.backward()
    print(x.grad)

>>> tensor([0.3333, 0.3333, 0.3333])
>>> tensor([0.3333, 0.3333, 0.3333])
>>> tensor([0.6667, 0.6667, 0.6667])
>>> tensor([0.6667, 0.6667, 0.6667])
>>> tensor([1., 1., 1.])
>>> tensor([1., 1., 1.])
>>> tensor([1.3333, 1.3333, 1.3333])
>>> tensor([1.3333, 1.3333, 1.3333])
>>> tensor([1.6667, 1.6667, 1.6667])

#Seems not. Gradients accumulate every second step :)

```
 
We can discard this PR and keep the optimizer_step logic to include zero_grad for the current optimizer only and solve the GAN training issue + we can enjoy the speedup from not calculating the unnecessary gradients. I'm going to submit another PR in a moment. Which one do you think would be a better solution?

**Comment by williamFalcon** [comment]
makes sense. let’s discard this one and use the second approach
